Step 2 of the interval design: Implement a threadsafe datastore

cnsnyder opened the pull request:
This is the implementation of step 2 of the interval design.

This step depends on all prior steps.

This PR adds a threadsafe datastore into the virt-who codebase. The datastore (as of this PR) is not in use in virt-who, it is to be used in subsequent steps of the design. More specifically, this datastore will be used to allow "destination" threads to consume reports produced by one or more "source" threads.

Related trello card: https://trello.com/c/HvuaQygd/275-interval-design-step-2-create-datastore
Design doc for the overall design: https://docs.google.com/document/d/1AV-eEp1HeemTHIiigGsUBDFK3W1XWMC1flSt9jrxek4/edit

cnsnyder commented on tests/test_datastore.py:
```diff
@@ -0,0 +1,224 @@
+        # places
+        return item, item
+
+    def assert_equivalent(self, item1, item2):
```
Self comment, this might be absurd. I could be convinced to trust that any object we'd like to test equivalence of has a sane __eq__ implementation method.

pcreech replied:
It can be considered rather complex, where we could test for standard equivalence.  Are we anticipating a need for storing complex custom objects?  Or standard python types and primitives?  if it's the latter, we can probably assume a sane "\_\_eq__" implementation

cnsnyder replied:
Khowell had a good idea that I think I'll go ahead and implement for this PR. That being to rely on the objects being put into the store having a 'copy' or 'clone' method that will return a copy of the object for storage.

Alternately, we could have the objects to be put in the store implement an interface that requires a serialize and deserialize method. We could serialize the objects on the way in to the store and deserialize them on the way out.

In either of the above case we can test that the implemented method on the objects being copied, serialized or deserialized produce suitable copies. Then all we need to see is that the datastore uses the right method on the way in and out.

cnsnyder commented:
@pcreech I've pushed a few updates to this PR. The new impl only allows "Cloneable" objects to be placed into the store and only returns a clone of the objects placed into the store.

The implementation of Cloneable.clone by default is to perform a deep copy of self.
I've added a few tests to show that the cloned objects behave the way we'd like.
I've removed a few tests (and reworked some) for the datastore.

The commits are still separate to make removing and modifying any or all of them easy.

Let me know if you like this version better than the original and I'll update the PR accordingly (squash if liked, remove / rework if not).

pcreech commented on virtwho/datastore.py:
```diff
@@ -0,0 +1,68 @@
+from threading import Lock
+
+
+
```
flake8: Too many lines

pcreech commented on virtwho/datastore.py:
```diff
@@ -0,0 +1,68 @@
+            try:
+                item = cPickle.loads(self._datastore[key])
+                return item
+            except KeyError as e:
```
flake8: Unused variable 'e'

SHould be able to omit 'as e' here

pcreech commented on virtwho/datastore.py:
```diff
@@ -0,0 +1,70 @@
+from threading import Lock
+
+
+
```
Flake8: Too many blank lines

pcreech commented on virtwho/datastore.py:
```diff
@@ -0,0 +1,70 @@
+            try:
+                item = pickle.loads(self._datastore[key])
+                return item
+            except KeyError as e:
```
Flake8: local variable 'e' is assigned but never used

except KeyError: should work here (tested while reviewing)

pcreech commented on tests/test_datastore.py:
```diff
@@ -0,0 +1,141 @@
+        self.mock_lock = mock_lock_instance
+        self.addCleanup(lock_patcher.stop)
+
+
```
flake8: too many blank lines

pcreech commented on tests/test_datastore.py:
```diff
@@ -0,0 +1,141 @@
+        test_item = "test_value"
+        test_key = "test_item"
+        datastore, mock_internal_ds = self.mock_test_data(Datastore(),
+                                                test_item=test_item)
```
flake8: under-indented for visual indent

pcreech commented:
LGTM: Thanks!
